emop-maven-plugin/src/main/java/edu/cornell/AffectedSpecsMojo.java

```diff
@@ -224,26 +228,73 @@ protected String getClassPath() throws MojoExecutionException {
         return Writer.pathToString(getSureFireClassPath().getClassPath());
     }
 
-    private void makeSourcesFile(String sourceList, Set<String> impacted) {
+    /**
+     * Given a path to a class file, returns a path to its corresponding source file. Assumes a standard directory
+     * layout, i.e., one where the source for {@code com.abc.A} resides in {@code sourceDir/com/abc/A.java}.
+     * @param classFile the path to the class file
+     * @param classesDir the base class file directory
+     * @param sourceDir the base sources directory
+     * @return the path to the source file
+     */
+    private static Path classFileToSource(Path classFile, Path classesDir, Path sourceDir) {
+        Path parent = sourceDir.resolve(classesDir.relativize(classFile)).getParent();
+        return parent.resolve(classFile.getFileName().toString().replace(".class", ".java"));
+    }
+
+    private void makeSourcesFile(String sourceList, Set<String> impacted) throws MojoExecutionException {
         Set<String> classes = new HashSet<>();
+        Path testSourceDir = getTestSourceDirectory().toPath().toAbsolutePath();
+        Path sourceDir = Paths.get(getTestSourceDirectory().getAbsolutePath().replace(
+                "src" + File.separator + "test", "src" + File.separator + "main"));
+
         for (String klas : impacted) {
             if (klas.contains("$")) {
                 klas = klas.substring(0, klas.indexOf("$"));
             }
             klas = klas.replace(".", File.separator) + ".java";
-            File test = new File(getTestSourceDirectory().getAbsolutePath() + File.separator + klas);
-            File source = new File(test.getAbsolutePath().replace("src" + File.separator + "test",
-                    "src" + File.separator + "main"));
+            File test = testSourceDir.resolve(klas).toFile();
+            File source = sourceDir.resolve(klas).toFile();
             if (source.exists()) {
                 classes.add(source.getAbsolutePath());
             } else if (test.exists()) {
                 classes.add(test.getAbsolutePath());
             } else {
                 getLog().error("Source file not found: " + source.getAbsolutePath());
                 getLog().error("Test file not found: " + test.getAbsolutePath());
+                throw new MojoExecutionException("Couldn't find source file for impacted class");
```

Comment:
Do we want to throw an entire exception in this case? How often does the code not find the source file? If it's a regular occurrence it may be better to just log the error and continue execution - similar question with the exceptions thrown later on in the function

Reply:
Hmm, it seems like if we fail to find the source file then that would be a serious issue and possibly a bug in RPS or STARTS. And yet, after checking the logs of past runs, it seems that this does occur quite frequently. So I will remove the exception for now, but we should probably look into this more. The exceptions later on in the function were as per @owolabileg's recommendation, so as to ensure that the path manipulations are being performed correctly.